plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -338,7 +363,77 @@ public IcebergTableHandle getTableHandle(
                 Optional.empty());
     }
 
-    private static long getSnapshotIdFromVersion(Table table, ConnectorTableVersion version)
+    private static Optional<TableMetadata> getInitialSnapshotMetadata(Table table, long snapshotId)
```

Comment:
This method looks very complicated. We need that for getting the schema and spec of the table as of given snapshot. I would expect this be doable with Iceberg table API.

@danielcweeks what's the recommended way to get these?

Reply:
I see in  https://github.com/trinodb/trino/pull/12786/commits/08cc0979fe02d93a24ed3310a16b32746d563844#r908684110
that `table.schemas().get(table.snapshot(snapshotId).schemaId())`  was attempted.
This looks much more right, but apparently didn't work. Maybe we need to understand more why it didn't?

Reply:
I was looking for a more concise way to get to the table metadata information, but haven't found any in iceberg.

I found then an approach for collecting table metadata files and built on top of this idea.

https://github.com/apache/iceberg/blob/b91bb9ce062b97efcb840008dce7552b83649e9a/core/src/main/java/org/apache/iceberg/ReachableFileUtil.java#L62

> `table.schemas().get(table.snapshot(snapshotId).schemaId())` this looks correct, but apparently it didn't work

This is indeed correct. 
The partition spec on the other hand is not available when starting from the snapshot and that's why I went through the metadata files to retrieve it.

The method `io.trino.plugin.iceberg.IcebergTableHandle#getPartitionSpecJson` is used mainly in `IcebergPageSink` for writing the data. If we'd abstain from transporting the partition spec within the `IcebergTableHandle`, we would not need the complex logic that this PR is introducing.
If however, there are good reasons to keep the partition spec in the `IcebergTableHandle`, I would argue then that we need this complex logic.

Reply:
> The partition spec on the other hand is not available when starting from the snapshot and that's why I went through the metadata files to retrieve it.

i think i saw @alexjo2144  pointing out we don't need partition spec.

Maybe we could just make it `Optional` in the table handle?

> The method `io.trino.plugin.iceberg.IcebergTableHandle#getPartitionSpecJson` is used mainly in `IcebergPageSink` for writing the data.

"mainly" or "only"?

Reply:
There's also a usage on `io.trino.plugin.iceberg.IcebergMetadata#getInfo` to figure out whether the table is partioned or not.

Reply:
> There's also a usage on `io.trino.plugin.iceberg.IcebergMetadata#getInfo` to figure out whether the table is partioned or not.

Good point.
We will need to make `io.trino.plugin.iceberg.IcebergInputInfo#partitioned` an `Optional<Boolean>` (as we don't want to provide fake info there)

cc @gaurav8297 @sopel39 cause i don't understand how https://github.com/trinodb/trino/pull/12551 is used

Reply:
I think it's okay to make it optional.